### PR TITLE
Fix/avoid multiple schedule deletions

### LINKED ISSUE #2569: Users sometimes delete the same schedule multiple times due to delay before the schedule disappears
### Issue Summary

It's possible to delete the same Schedule multiple times if the connection is slow enough - there is nothing to visually show something is happening.

Schedule needs the following popup to be implemented, when a delete is in progress the "Delete Schedule" button should show the loading spinner to prevent further confusion/actions.

⚠️ As part of this fix the same loading spinner should be shown during "Disband Team" for completionism sake.

image

### Steps to Reproduce

1. Press more actions on the Schedule
2. Press delete Schedule - the dropdown will close
3. Open the dropdown again
4. Press delete Schedule again.

### PR DESCRIPTION
**Pull request by arthur1041**
## What does this PR do?

Prevents users from deleting the same schedule multiple times due to delay before the schedule disappears. It also applies the same fix to team disband.
(Also fixes a unique key error that was not in the [issue](https://github.com/calcom/cal.com/issues/2569) at the "teams" page)

### New behaving:
Schedule deletion:
schedule_deletion_new_behaving

Team disband:
team_disband_new_behaving

Fixes [#2569](https://github.com/calcom/cal.com/issues/2569)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- [X] In order to check the solution it is only needed to run the application and try to remove a schedule or disband a team.
- [X] This problem only occurs with poor internet conection. If your internet connection is good enough, you might need to use a setTimeout involving "deleteFunction" at apps/web/components/availability/ScheduleListItem.tsx and involving "props.onActionSelect("disband")" at apps/web/components/team/TeamListItem.tsx

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by emrysal — DISMISSED** [review]
Thanks for this, I'm tagging @Jaibles to have his voice heard on the design decisions made in this PR - from a code perspective all looks good, but one recommendation:

You're adding a useState() to track isDeleting, ideally what I'd like to see is the addition of an isDeleting prop to TeamListItem, in the TeamList (not part of this PR) you can then use this new prop like `isDeleting={deleteTeamMutation.isLoading}` as documented @ https://trpc.io/docs/react-mutations. 

Edit: Thinking more about this this does not work quite that way - given the deleteTeamMutation is item specific and not list specific. To tackle this an alternative is to move the deleteTeamMutation to the TeamListItem with callback to the higher level TeamList to ensure the list is revalidated post-deletion.

**Comment by arthur1041** [comment]
> Thanks for this, I'm tagging @Jaibles to have his voice heard on the design decisions made in this PR - from a code perspective all looks good, but one recommendation:
> 
> You're adding a useState() to track isDeleting, ideally what I'd like to see is the addition of an isDeleting prop to TeamListItem, in the TeamList (not part of this PR) you can then use this new prop like `isDeleting={deleteTeamMutation.isLoading}` as documented @ https://trpc.io/docs/react-mutations.
> 
> Edit: Thinking more about this this does not work quite that way - given the deleteTeamMutation is item specific and not list specific. To tackle this an alternative is to move the deleteTeamMutation to the TeamListItem with callback to the higher level TeamList to ensure the list is revalidated post-deletion.

Hey! Just added a new commit. Can you check if my modifications correspond to your suggestion?

**Comment by arthur1041** [comment]
You might need to activate network throttling in your browser to simulate a slow connection and see the loading icons.

**Comment by emrysal** [comment]
I've caught with @Jaibles and he reckons not to introduce an alternative loading style, instead recommending to using the Button `loading` parameter, both for simplicity and consistency sake.

**Comment by Jaibles** [comment]
@arthur1041 - looks good. My only comment on design would be to use the button loading state that we commonly use on the button and on the dropdown. Right now it's currently using the app loader.

[The button designs are here.](https://www.figma.com/file/MBWYmOwg7esjrcNeyDVYty/%F0%9F%94%98-Cal-DS?node-id=550%3A18864)
CleanShot 2022-05-04 at 12 16 14@2x

Let me know if you have any questions!

**Comment by arthur1041** [comment]
> y comment on design would be to use the button loading state as @emrysal mentioned Looks like he's probably provided what you needed but the designs are [here](https://www.figma.com/file/MBWYmOwg7esjrcNeyDVYty/%F0%9F%94%98-Cal-DS?node-id=550%3A18864) in c

Got it. I'll add this modification.

File: apps/web/components/availability/ScheduleListItem.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+import { availabilityAsString } from "@calcom/lib/availability";
+import { useLocale } from "@calcom/lib/hooks/useLocale";
+import { Availability } from "@calcom/prisma/client";
+import { Button } from "@calcom/ui";
+import Dropdown, { DropdownMenuContent, DropdownMenuItem, DropdownMenuTrigger } from "@calcom/ui/Dropdown";
+
+import Loader from "@components/Loader";
+
+export function ScheduleListItem({ schedule, deleteFunction }: { schedule: any; deleteFunction: Function }) {
```
Comment (zomars) [inline apps/web/components/availability/ScheduleListItem.tsx:13]: We try avoid `any`s unless strictly necessary.

File: apps/web/components/team/TeamListItem.tsx [hunk 2]
```diff
@@ -33,7 +34,7 @@ import { TeamRole } from "./TeamPill";
 interface Props {
   team: inferQueryOutput<"viewer.teams.list">[number];
   key: number;
-  onActionSelect: (text: string) => void;
+  deleteTeamMutation: UseMutationResult<any, any, any, any>;
```
Comment (zomars) [inline apps/web/components/team/TeamListItem.tsx:37]: This is a no-no let's figure it out how to make this type safe.

File: apps/web/components/team/TeamListItem.tsx [hunk 3]
```diff
@@ -175,7 +188,10 @@ export default function TeamListItem(props: Props) {
                           variety="danger"
                           title={t("disband_team")}
                           confirmBtnText={t("confirm_disband_team")}
-                          onConfirm={() => props.onActionSelect("disband")}>
```
Comment (zomars) [inline apps/web/components/team/TeamListItem.tsx:178]: What's the reason for this change? AFAIK is not adding anything.

**Review by zomars — CHANGES_REQUESTED** [review]

**Review by zomars — APPROVED** [review]

File: apps/web/components/availability/ScheduleListItem.tsx (hunk 1 above)
Comment (arthur1041) [reply apps/web/components/availability/ScheduleListItem.tsx]: I didn't find the correct type at the moment so I use `any`. I saw you already committed an correction.

File: apps/web/components/team/TeamListItem.tsx (hunk 3 above)
Comment (arthur1041) [reply apps/web/components/team/TeamListItem.tsx:178]: It was part of the modifications I did for @emrysal's changes request.

Comment (zomars) [reply apps/web/components/team/TeamListItem.tsx:178]: I think what @emrysal suggested was to remove the need for an extra useState. Which could be solved by just passing the mutation isLoading property as a prop.

**Pull request merged**